### DISCUSSION TITLE
docker内でgazebo起動時にハードウェアアクセラレーションする方法について

### DISCUSSION THREAD

**Discussion by seigot** [Tiryoh/docker-ros-desktop-vnc#24] [General]
docker_ros-desktop-vnc docker内でgazebo起動時にハードウェアアクセラレーションする方法について質問があります。

通常、docker内ではgpuは使えないと認識しています。そのため、docker内でgazeboを使うと、docker外でgazeboを使う場合と比べてrealtimefatorが低く出る傾向があると認識しています。
nvidia dockerを使えばこれが解決できるかもと考えているのですが、何かご存知ありませんでしょうか。。

**Comment by Tiryoh** [comment]
> nvidia dockerを使えばこれが解決できるかもと考えているのですが、何かご存知ありませんでしょうか。。

ハードウェアアクセラレーションについて考えたことはありませんでした。確認してみます。

**Comment by Tiryoh** [comment]
とりあえず思いつくのは、このDockerイメージでベースとしている[`dorowu/ubuntu-desktop-lxde-vnc`](https://hub.docker.com/r/dorowu/ubuntu-desktop-lxde-vnc/)のベースを、NVIDIA Container Toolkitに対応したGPUを使用するイメージに置き換えることです。
とりあえず[`tiryoh/ubuntu-desktop-lxde-vnc:bionic-cuda-10.2`](https://hub.docker.com/r/tiryoh/ubuntu-desktop-lxde-vnc/tags?page=1&name=melodic-cuda)（[Dockefile](https://github.com/Tiryoh/docker-ubuntu-vnc-desktop/blob/bionic-cuda-10.2/Dockerfile.amd64)）として[`nvidia/cuda`](https://hub.docker.com/r/nvidia/cuda/)ベースのCUDA対応版`dorowu/ubuntu-desktop-lxde-vnc`を作りました。
そして、とりあえず[`tiryoh/ros-desktop-vnc:melodic-cuda-10.2`](https://hub.docker.com/r/tiryoh/ros-desktop-vnc/tags?page=1&name=melodic-cuda)（[Dockerfile](https://github.com/Tiryoh/docker_ros-desktop-vnc/blob/feature/support-cuda/melodic-cuda-10.2/Dockerfile)）として`tiryoh/ubuntu-desktop-lxde-vnc-cuda`ベースの`tiryoh/ros-desktop-vnc`を作ってみました。
一応手元のMX150入りのノートパソコンでは動いているようですが、GPUを有効にできたことをどうやって確認したらよいのか、というところでちょっと悩んでいます。

**Comment by Tiryoh** [comment]
> 一応手元のMX150入りのノートパソコンでは動いているようですが、GPUを有効にできたことをどうやって確認したらよいのか、というところでちょっと悩んでいます。

この件について、Twitterにて[inata39](https://twitter.com/inata39)さんより以下の[アドバイスをいただきました](https://twitter.com/inata39/status/1297475986028523525?s=20)。

> Docker上でシミュレーターを起動させ、「$ nvidia-smi l」コマンドでGPUが受け持っているprocessが表示されるので、それっぽいプロセスがいるか？
を確認する
> というのはどうでしょう

確認してみます。

**Comment by Tiryoh** [comment]
以下のコマンドでNVIDIA Container Toolkitを利用したコンテナを起動してみました。

```
$ docker run --gpus all -p 6080:80 --shm-size=512m --rm tiryoh/ros-desktop-vnc:melodic-cuda-10.2
```

デスクトップ環境がGPUを利用できてないようで、X11やgzclientが`nvidia-smi`コマンドの出力結果に出てきませんでした。
スクリーンショットの撮影方法がよくなかったのでちょっと見づらいですが、`Processes:`のところにプロセス名は何も出ていません。

![Screenshot from 2020-08-23 20-14-17](https://user-images.githubusercontent.com/3256629/90977052-65370a00-e57d-11ea-8f01-113fef506364.png)

参考までに以下がGPUが有効になっているPC上で（Docker上ではなく）、Gazeboを起動した際の`nvidia-smi`コマンドの実行結果です。

GazeboがGPUを利用できているので`gzserver`と`gzclient`がプロセス名として出てきます。
また、デスクトップ環境がGPUを利用しているので`/usr/bin/gnome-shell`や`/usr/lib/xorg/Xorg`が同様にプロセス名として列挙されていることを確認できます。

```
$ nvidia-smi
Sun Aug 23 20:08:36 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.100      Driver Version: 440.100      CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce MX150       Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   61C    P0    N/A /  N/A |   1578MiB /  2002MiB |     24%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      2002      G   /usr/lib/xorg/Xorg                           626MiB |
|    0      2171      G   /usr/bin/gnome-shell                         261MiB |
|    0     15995      G   ...AAAAAAAAAAAACAAAAAAAAAA= --shared-files   414MiB |
|    0     26858      G   gzclient                                      90MiB |
|    0     26871      G   gzserver                                       3MiB |
|    0     28418      G   ...quest-channel-token=9490849909445542573   177MiB |
+-----------------------------------------------------------------------------+
```

**Comment by seigot** [comment]
回答が遅れまして申し訳ありません。上記お試し頂きありがとうございます。
確認なのですが、上記は以下どちらの理解とすればよろしいでしょうか。
　A. そもそも、「VIDIA Container Toolkitに対応したGPUを使用するイメージ」でも、GPUを有効にできていない（nvidia-smiコマンドの出力結果に出てこない）
　B.「VIDIA Container Toolkitに対応したGPUを使用するイメージ」ではGPUを有効にできるが、今回試行頂いたnvidia/cudaベースのCUDA対応版では、GPUを有効にできていない

また、本件について色々調べていたところ以下記事を見つけました。直接のLinuxPC環境ではないですが、何か私の方でも確認できましたら共有させて頂きます。
https://qiita.com/yamamo-to/items/51bfb92a00033257f02a

**Comment by seigot** [comment]
以下を参考にしてnvidia-dockerをインストールした後、
https://medium.com/nvidiajapan/nvidia-docker-%E3%81%A3%E3%81%A6%E4%BB%8A%E3%81%A9%E3%81%86%E3%81%AA%E3%81%A3%E3%81%A6%E3%82%8B%E3%81%AE-20-09-%E7%89%88-558fae883f44

同サイトの「GPU 付きコンテナの起動方法」を試しましたが、
nvidia-smiコマンド実行時にプロセスが出力されませんでした。
> # すべての GPU を使用
> docker run --gpus all --rm nvidia/cuda:9.0-base nvidia-smi

Tiryohさんに用意頂いたコマンドでも同様です。

```
$ docker run --gpus all -p 6080:80 --shm-size=512m --rm tiryoh/ros-desktop-vnc:melodic-cuda-10.2
...
# nvidia-smi 
Sat Sep 19 06:24:49 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.51.06    Driver Version: 450.51.06    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 2060    On   | 00000000:01:00.0  On |                  N/A |
| N/A   40C    P8     3W /  N/A |    261MiB /  5932MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+

```

他サイトをみると、「コンテナの中でnvidia-smiが実行できたら成功」みたいな表現はあるので、
GPU自体は使えるがアプリ側の対応が要るのかな？と考えています。
http://verifiedby.me/adiary/0137

更に他サイトをみると、
「NVIDIA_VISIBLE_DEVICES を指定しただけでは、nvidia-smi コマンドは使えても OpenGL や CUDA を利用することができない。
これらを利用するためには、用途に応じて環境変数 NVIDIA_DRIVER_CAPABILITIES を指定する必要がある。」
みたいな記述もあるので色々試したいと考えています。（手詰まってきてる感はあります）
https://qiita.com/hoto17296/items/be34849412e7ae9db6ff



ちなみに、用意頂いたコマンドを実行すると、ブラウザからlocalhost:6080 で入れなくなるのですが何かご存知でしょうか、、？
$ docker run --gpus all -p 6080:80 --shm-size=512m --rm tiryoh/ros-desktop-vnc:melodic-cuda-10.2

**Comment by Tiryoh** [comment]
バタバタしててお返事できてませんでした

> 確認なのですが、上記は以下どちらの理解とすればよろしいでしょうか。
> 　A. そもそも、「VIDIA Container Toolkitに対応したGPUを使用するイメージ」でも、GPUを有効にできていない（nvidia-smiコマンドの出力結果に出てこない）
> 　B.「VIDIA Container Toolkitに対応したGPUを使用するイメージ」ではGPUを有効にできるが、今回試行頂いたnvidia/cudaベースのCUDA対応版では、GPUを有効にできていない

上記の質問、解決したようですが、記録のためコメントしておきます。
ABどちらでもなく、GPU自体は使えているものの、使いたいアプリケーションがGPUを利用せずCPUを利用しているという状況です。
具体的には

* DockerコンテナをホストしているPC ← GPU利用可能
* nvidia/cudaベースのCUDA対応版dorowu/ubuntu-desktop-lxde-vnc (tiryoh/ubuntu-desktop-lxde-vnc-cuda) ← GPU利用可能、ただしデスクトップアプリケーションがGPUを利用していない
* tiryoh/ubuntu-desktop-lxde-vnc-cudaベースのtiryoh/ros-desktop-vnc ← GPU利用可能、ただしデスクトップアプリケーションがGPUを利用していない

という状況です。

> 他サイトをみると、「コンテナの中でnvidia-smiが実行できたら成功」みたいな表現はあるので、
GPU自体は使えるがアプリ側の対応が要るのかな？と考えています。

はい、同じように考えています。
[先のコメント](https://github.com/Tiryoh/docker-ros-desktop-vnc/issues/12#issuecomment-678761919)にて書いたように、デスクトップ (x11)をGPU対応させることが次にためせることだと思われますが、現在取りかかれていない状況です。

> ちなみに、用意頂いたコマンドを実行すると、ブラウザからlocalhost:6080 で入れなくなるのですが何かご存知でしょうか、、？
> $ docker run --gpus all -p 6080:80 --shm-size=512m --rm tiryoh/ros-desktop-vnc:melodic-cuda-10.2

こちらでは特にWebブラウザからアクセスできなくなることはありませんでした。何かエラー等は出ていますか？
すでに6080番ポートを別のコンテナが使用していると、アクセスできなくなるとは思います。

**Comment by Tiryoh** [comment]
ちょっと時間があいてしまいましたが、GPUを利用したイメージの作成について調べてみました。

[Xvfb](http://manpages.ubuntu.com/manpages/precise/ja/man1/Xvfb.1.html)でデスクトップアプリケーションを起動していますが、これがGPUを利用できていないようです。
Xvfbを呼び出している場所：https://github.com/Tiryoh/docker-ubuntu-vnc-desktop/blob/07244b4af28044e267c9abf752d9b217b5e0b547/rootfs/etc/supervisor/conf.d/supervisord.conf#L50

普段利用しているデスクトップアプリケーションとは別に、Docker内でGPUを利用してデスクトップアプリケーションを実行する例はあまりないようです。
多くはX11転送を使ってGUIのアプリケーションを起動しているようです。この方法はこのDockerfileで実現したい「Webブラウザ内で操作が完結するデスクトップ環境」とは異なります。
少し調べてみると以下の資料にたどり着きました。

* https://medium.com/@trn84/running-opengl-cuda-applications-with-nvidia-docker2-on-a-remote-headless-gpu-system-6b19c665286d
* https://github.com/ryought/glx-docker-headless-gpu

[ryought/glx-docker-headless-gpu](https://github.com/ryought/glx-docker-headless-gpu)や[willkessler/nvidia-docker-novnc](https://github.com/willkessler/nvidia-docker-novnc)のようにGCPやAWS上の仮想マシンで実行している例はあるようですが、これはそのまま手元のマシンで実行するには不向きでした。
コードを読めばなんとなくわかりますが、気になってつい実行してみたので後のために記録しておくと、
実行すると普段利用しているデスクトップアプリケーションであるGNOMEではなく、Docker内部のデスクトップがモニタに表示されるようになってしまいました。
キーボードもモニタも繋がないクラウド上の仮想マシンならこれでよいですが、普段のPC上で使用するのには向いていませんでした。

`--allow-empty-initial-configuration`をつけて`nvidia-xconfig`を実行するとよいようですが、2020年10月17日現在、[nvidia/cudagl](https://hub.docker.com/r/nvidia/cudagl)には`nvidia-xconfig`が入っていないようでした。

https://7me.nobiki.com/2019/11/11/xorg-virtual-screen-with-nvidia-gpu/

そのため[ryought/glx-docker-headless-gpuのDockerfile#L54-L77](https://github.com/ryought/glx-docker-headless-gpu/blob/85167001e81167444eb1ae4cea4cc03cb281fc26/Dockerfile#L54-L77)のようにDockerfile内でnvidia-installerを実行する必要がありそうで、汎用的に使うのは難しそうです。

また機会があれば調べてみますが、難しそうというのがわかったところでこの件は一旦保留にしたいと思います。

**Comment by seigot** [comment]
すみません、丁寧に回答頂きありがとうございます。
せっかく調べてまとめて頂いたのですが、私の知識では2割未満の理解度です。。
汎用的な解決はないとの事で、使うデバイスを統一したり、AWS等クラウドの場合は運用ルールに準拠するといった配慮が要りそうですね。
（細かいオプションは理解できていないです） 

また、docker内のGPUを利用してる方はもっと多いと思っていたのですがあまり事例がないのですね。
docker内GPU有無で処理速度がどう変わるかベンチマークとってる方も見かけないです。

また何か分かりましたらアップデートさせて頂きますが、一旦保留で問題ございません。

**Comment by Tiryoh** [comment]
レスポンスありがとうございます。ご質問頂いて私も気になったので調べた結果をメモしてみました。
もう少しわかりやすく説明できればよかったのですが、理解が曖昧なところも多くて説明が難しく、歯がゆいです……

ツールを色々組み合わせているからだと思いますが、私もそれぞれのツールの使い方をなんとなくわかった程度です。
探せばもっといろんなオプションやもっと賢いやり方があるかもしれませんし、どこかで誰かの参考になるかもしれないので、とりあえずこのissueはこのまま残しておきます。

**Comment by Jay13522** [comment]
@Tiryoh Thanks for building a great docker image.
But I have the same problem recently. When I launch a heavy project with Gazebo, it runs slowly. Because gzclient and gzserver can't run with GPU.
So have you solve this problem now?

**Reply by Tiryoh** [reply]
Hi @Jay13522, Thanks for the comment! But unfortunately, it has not been resolved.

This is the current status.

I'm planning to try this Issue again soon as I'm able to run it using the GPU with [docker-ros-melodic-desktop](https://github.com/Tiryoh/docker-ros-melodic-desktop) (which is not the one used from a web browser) and [osrf/rocker](https://github.com/osrf/rocker).

## What has been achieved

* access GPU of the host machine from inside the docker container
    * [`tiryoh/ubuntu-desktop-lxde-vnc:bionic-cuda-10.2`](https://hub.docker.com/r/tiryoh/ubuntu-desktop-lxde-vnc/tags?page=1&name=melodic-cuda) ([Dockefile](https://github.com/Tiryoh/docker-ubuntu-vnc-desktop/blob/bionic-cuda-10.2/Dockerfile.amd64)) and [`tiryoh/ros-desktop-vnc:melodic-cuda-10.2`](https://hub.docker.com/r/tiryoh/ros-desktop-vnc/tags?page=1&name=melodic-cuda) ([Dockerfile](https://github.com/Tiryoh/docker_ros-desktop-vnc/blob/feature/support-cuda/melodic-cuda-10.2/Dockerfile))
        * based on [`nvidia/cuda`](https://hub.docker.com/r/nvidia/cuda/)

## What has not been achieved

* launch lxpanel (desktop application) from [Xvfb](http://manpages.ubuntu.com/manpages/precise/ja/man1/Xvfb.1.html) using GPU from inside the docker container
* launch Gazebo using GPU from inside the docker container

**Comment by Tiryoh** [comment]
https://github.com/atinfinity/nvidia-egl-desktop-ros2

こちらのDockerfileでGPUを利用してGazeboを実行できることを確認しました。

https://github.com/ehfd/docker-nvidia-egl-desktop がベースになっているようです。

**Comment by ffunatsu** [comment]
素晴らしいDockerイメージを作っていただき、本当に感謝致します。

動作報告になりますが、/etc/apt/sources.listのミラーがうまく動かないなどの問題があったので、下記のようなDockerfileを作成し、CondaのPyTorch環境との共存と、CUDA-10.2のincludeの利用やnvccのインストールなども問題なくできました。

`COPY cuda-10.2` で使うcuda-10.2フォルダの内容は、`/usr/local/cuda`の中身のincludeとnv*で、下記Docker内で`sudo apt-get install -y cuda-10-2`したときの`/usr/local/cuda`をバックアップしたもので、includeとnvvmなどが不要なプロジェクトの場合はコメントアウトしても問題なく動作します。（Dockerfile内でそのまま行うと、library/driverミスマッチが起こるので、あえて必要なものだけコピーして使ってます。）

```Dockerfile
FROM tiryoh/ros-desktop-vnc:melodic-cuda-10.2

COPY sources.list /etc/apt/sources.list
RUN sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
RUN curl -s https://raw.githubusercontent.com/ros/rosdistro/master/ros.asc | sudo apt-key add -
RUN rm /etc/apt/sources.list.d/cuda*
RUN rm /etc/apt/sources.list.d/nvidia*
RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin
RUN mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600
RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/7fa2af80.pub
RUN apt-key del 7fa2af80
RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-keyring_1.0-1_all.deb
RUN dpkg -i cuda-keyring_1.0-1_all.deb
RUN add-apt-repository "deb https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/ /"
RUN apt update
WORKDIR /opt
RUN wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh && \
    sh Miniconda3-latest-Linux-x86_64.sh -b -p /opt/miniconda3 && \
    rm -r Miniconda3-latest-Linux-x86_64.sh
ENV PATH $PATH:/opt/miniconda3/bin
RUN conda create -n py38 python=3.8.16
RUN conda init bash
COPY cuda-10.2/include /usr/local/cuda-10.2/include
RUN bash -c "source activate py38 && conda install pytorch==1.9.1 torchvision==0.10.1 torchaudio==0.9.1 -c pytorch"
COPY cuda-10.2/bin /usr/local/cuda-10.2/bin
COPY cuda-10.2/nvml /usr/local/cuda-10.2/nvml
COPY cuda-10.2/nvvm /usr/local/cuda-10.2/nvvm
COPY cuda-10.2/nvvmx /usr/local/cuda-10.2/nvvmx
RUN chmod 777 /opt/miniconda3/envs/py38/lib/python3.8/site-packages/

RUN bash -c "conda config --set auto_activate_base false"
ENV PATH $PATH:/usr/local/cuda/nvvm/bin
```

`COPY sources.list /etc/apt/sources.list` での  sources.list は以下の通りです。

```
deb http://jp.archive.ubuntu.com/ubuntu/ bionic main restricted
deb http://jp.archive.ubuntu.com/ubuntu/ bionic-updates main restricted
deb http://jp.archive.ubuntu.com/ubuntu/ bionic universe
deb http://jp.archive.ubuntu.com/ubuntu/ bionic-updates universe
deb http://jp.archive.ubuntu.com/ubuntu/ bionic multiverse
deb http://jp.archive.ubuntu.com/ubuntu/ bionic-updates multiverse
deb http://jp.archive.ubuntu.com/ubuntu/ bionic-backports main restricted universe multiverse
deb http://security.ubuntu.com/ubuntu bionic-security main restricted
deb http://security.ubuntu.com/ubuntu bionic-security universe
deb http://security.ubuntu.com/ubuntu bionic-security multiverse
```

**Reply by Tiryoh** [reply]
情報ありがとうございます！Miniconda3を使用するのですね:eyes:
